### ISSUE TITLE
Reduced write performance with every additional index by factor ~2

### ISSUE THREAD

**Issue by sandrokeil** [arangodb/arangodb#7703]
## My Environment

* __ArangoDB Version__:       3.4.0
* __Storage Engine__:            RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Manual Start in Docker
* __Configuration__:               default
* __Infrastructure__:               local via Docker (18.09.0-ce with ovlerlay2 and Native Overlay Diff: true)
* __Operating System__:        From [ArangoDB Docker Image](https://hub.docker.com/r/arangodb/arangodb/)
* __Used Package__:              Docker - official Docker library

## Component, Query & Data

__Affected feature__:
RocksDB write performance with multiple indexes.

## Steps to reproduce

1. Create Collection via arangosh
```
db._create('myCollection', {
    'keyOptions': {
        'allowUserKeys': false,
        'type': 'autoincrement',
        'increment': 1,
        'offset': '100000000000000000',
    }
});
```
2. Insert some documents with query profiler.
```
db._profileQuery(`
FOR i IN 1..1000
  INSERT { 
    name: CONCAT("test", i), 
    value: 100 + FLOOR(RAND() * (150 - 100 + 1)) 
  } IN myCollection
`);
```
3. Add an index and do step 2
```
db.myCollection.ensureIndex({ type: "skiplist", fields: [ "_key" ], unique: true });
```
4. Add an index and do step 2
```
db.myCollection.ensureIndex({ type: "hash", fields: [ "name" ]});
```

__Problem__:
Reduced write performance with every additional index by factor ~ 1,5 - 3. Almost same behavior if collection is created with all indexes and only one insert run. I have also a drastically write performance drawbacks in my benchmark suite if I add indexes.

- Execution time without index: 0.01620
- Execution time with 1 index: 0.04259
- Execution time with 2 index: 0.06600

__Expected result__:
Not that much performance drawback or is this a normal behavior?

**Comment by sandrokeil** [comment]
I just tested today ArangoDB devel vs. ArangoDB 3.4.2 with our [benchmark suite](https://github.com/prooph/event-store-benchmarks/pull/16) and the devel results are very impressive.

**3.4.2**
```
arangodb-vpack - Aggregate real world test duration 37 seconds
arangodb-vpack - Aggregate avg writes 337 events/second
arangodb-vpack - Aggregate avg reads 675 events/second
```

**devel**
```
arangodb-vpack - Aggregate real world test duration 3 seconds
arangodb-vpack - Aggregate avg writes 4166 events/second
arangodb-vpack - Aggregate avg reads 8333 events/second
```

Maybe this issue can be closed as soon as 3.5.0 is released? Is there a timeframe? What was the issue for "slow" performance?

**Comment by jsteemann** [comment]
@sandrokeil: thanks a lot for benchmarking and sharing the results here!

The numbers from devel indeed look very appealing :smile:!
I know that we optimized in devel a bit by not using a rocksdb::WriteBatchWithIndex (but just a simple rocksdb::WriteBatch) when populating new non-unique indexes. This improves performance quite a bit, but I would not expect this change to affect performance _after_ indexes were set up initially.
And I think the benchmark is measuring reads and write after the initial structure has been set up, right?

Just to double-check, did you use the "official" docker containers for the comparison, or did you build any of the versions manually? Just wondering because when compiling manually there may be different (more aggressive but less portable) optimization settings than when using the official releases.

Additionally just wondering whether the two versions were tested in a "clean state", i.e. the operations and measurements were carried out in an identical setup and database state.

And finally, is there any information about how many threads/concurrency the benchmark is using? Is it somehow sending multiple requests to the database in parallel or is it a single-threaded benchmark?

Thanks!

**Comment by sandrokeil** [comment]
I use the official [ArangoDB Preview Docker Image](https://hub.docker.com/r/arangodb/arangodb-preview) and [ArangoDB Docker Image](https://hub.docker.com/_/arangodb) with default config.

> I know that we optimized in devel a bit by not using a rocksdb::WriteBatchWithIndex (but just a simple rocksdb::WriteBatch) when populating new non-unique indexes. 

I found this only, because I had a bug in the index creation and was happy that it was so fast. :smile:  Depending on the stream strategy (*Aggregate* is the default) there are [up to 4 indexes](https://github.com/sandrokeil/arangodb-event-store/blob/master/src/PersistenceStrategy/AggregateStreamStrategy.php#L55), but they are unique. :thinking: 

### Benchmark

The real world benchmark contains:
- 50 processes are writing 250 events at the same time
- 6 projections are reading events at the same time
- a total of 12500 events are written
- a total of 15000 events are read

Indexes are created for each collection. Here we have at the moment `50 (processes) * 50 (collections) * 5 (documents)`. You can [customize it here](https://github.com/sandrokeil/event-store-benchmarks/blob/feature/arangodb-vpack/src/writer.php#L29). Let me know if I should test another test setup e.g. with more documents / readers / writers.

> Additionally just wondering whether the two versions were tested in a "clean state", i.e. the operations and measurements were carried out in an identical setup and database state.

Yes, each benchmark run has a clean state. I have some fluctuations of ~1s depending on my CPU turbo mode, I guess, but the results are reproducible. 

The benchmark system:
- AMD Threadripper 1950x
- 128 GB RAM @ 2933MHz
- Samsung SSD 960 Pro SSD (PCI-E)

*arangodb/arangodb-preview:devel* (best, worst is ~3.2)
```
arangodb - Aggregate real world test duration 2.18177873 seconds
arangodb - Aggregate avg writes 5729.27026381 events/second
arangodb - Aggregate avg reads 6875.12431657 events/second
```

*arangodb/arangodb-preview:3.4.2*
```
arangodb - Aggregate real world test duration 35.70852038 seconds
arangodb - Aggregate avg writes 350.05650940 events/second
arangodb - Aggregate avg reads 420.06781128 events/second
```

*arangodb:3.4.1*
```
arangodb - Aggregate real world test duration 35.70934944 seconds
arangodb - Aggregate avg writes 350.04838218 events/second
arangodb - Aggregate avg reads 420.05805861 events/second
```

I get this warning **only** with 3.4.1.
```
database_1  | automatically choosing storage engine
database_1  | Initializing database...Hang on...
database_1  | Database initialized...Starting System...
database_1  | 2019-01-21T21:18:48Z [1] INFO ArangoDB 3.4.1 [linux] 64bit, using jemalloc, build tags/v3.4.1-0-g1498c08084, VPack 0.1.33, RocksDB 5.16.0, ICU 58.1, V8 5.7.492.77, OpenSSL 1.1.0h  27 Mar 2018
database_1  | 2019-01-21T21:18:48Z [1] INFO detected operating system: Linux version 4.20.1-arch1-1-ARCH (builduser@heftig-17587) (gcc version 8.2.1 20181127 (GCC)) #1 SMP PREEMPT Wed Jan 9 20:25:43 UTC 2019
database_1  | 2019-01-21T21:18:48Z [1] WARNING {memory} maximum number of memory mappings per process is 65530, which seems too low. it is recommended to set it to at least 1024000
database_1  | 2019-01-21T21:18:48Z [1] WARNING {memory} execute 'sudo sysctl -w "vm.max_map_count=1024000"'
database_1  | 2019-01-21T21:18:48Z [1] INFO {authentication} Jwt secret not specified, generating...
database_1  | 2019-01-21T21:18:48Z [1] INFO using storage engine rocksdb
database_1  | 2019-01-21T21:18:48Z [1] INFO {cluster} Starting up with role SINGLE
database_1  | 2019-01-21T21:18:48Z [1] INFO {syscall} file-descriptors (nofiles) hard limit is 1048576, soft limit is 1048576
database_1  | 2019-01-21T21:18:48Z [1] INFO {authentication} Authentication is turned off, authentication for unix sockets is turned on
database_1  | 2019-01-21T21:18:49Z [1] INFO using endpoint 'http+tcp://0.0.0.0:8529' for non-encrypted requests
database_1  | 2019-01-21T21:18:49Z [1] INFO ArangoDB (version 3.4.1 [linux]) is ready for business. Have fun!
```

### Testing
You can test it with [my branch](https://github.com/sandrokeil/event-store-benchmarks/tree/feature/arangodb-vpack):

1. `git clone --branch feature/arangodb-vpack https://github.com/sandrokeil/event-store-benchmarks.git event-store-benchmark`
1. `docker run --rm -i -v $(pwd):/app prooph/composer:7.2 update -o`
1. `. bench_docker.sh --driver arangodb` (don't forget the leading dot)

To use the current stable release change [this line](https://github.com/sandrokeil/event-store-benchmarks/blob/feature/arangodb-vpack/docker-compose-arangodb.yml#L21) with the image you want to test and execute step 3 from above. There is also an experimental vpack (`. bench_docker.sh --driver arangodb-vpack`) test.

**Comment by dothebart** [comment]
Hi, 
thanks again for sharing your results and your benchmark suite. 
Closing this as solved with the availability of [ArangoDB 3.7](https://www.arangodb.com/2020/08/arangodb-3-7-a-big-step-forward-for-multi-model/).
